test/jdk/java/net/ServerSocket/ImplAccept.java

```diff
@@ -0,0 +1,293 @@
+    @Test
+    public void testClosed() throws Exception {
+        Socket socket = new Socket();
+        socket.close();
+
+        try (ServerSocket ss = serverSocketToAccept(socket);
+             Socket peer = new Socket(ss.getInetAddress(), ss.getLocalPort())) {
+
+            SocketImpl si = getSocketImpl(socket);
+
+            // accept should throw and peer should read EOF
+            assertThrows(IOException.class, ss::accept);
+            assertTrue(peer.getInputStream().read() == -1);
+
+            // the SocketImpl should have not changed
+            assertTrue(getSocketImpl(socket) == si);
+        }
+    }
+
+    /**
+     * Returns the socket's SocketImpl.
+     */
+    private static SocketImpl getSocketImpl(Socket s) {
+        try {
+            Field f = Socket.class.getDeclaredField("impl");
+            f.setAccessible(true);
+            return (SocketImpl) f.get(s);
+        } catch (Exception e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    /**
+     * Returns the SocketImpl that the given SocketImpl delegates to.
+     */
+    private static SocketImpl getDelegate(SocketImpl si) {
+        try {
+            Class<?> clazz = Class.forName("java.net.DelegatingSocketImpl");
+            Field f = clazz.getDeclaredField("delegate");
+            f.setAccessible(true);
+            return (SocketImpl) f.get(si);
+        } catch (Exception e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    /**
+     * Returns true if the SocketImpl is a DelegatingSocketImpl.
+     */
+    private static boolean isDelegatingSocketImpl(SocketImpl si) {
+        try {
+            Class<?> clazz = Class.forName("java.net.DelegatingSocketImpl");
+            return clazz.isInstance(si);
+        } catch (Exception e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    /**
+     * Returns true if the SocketImpl is a PlatformSocketImpl.
+     */
+    private static boolean isPlatformSocketImpl(SocketImpl si) {
+        try {
+            Class<?> clazz = Class.forName("sun.net.PlatformSocketImpl");
+            return clazz.isInstance(si);
+        } catch (Exception e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    /**
+     * Returns true if the SocketImpl has an open socket.
+     */
+    private static boolean isSocketOpen(SocketImpl si) throws Exception {
+        assertTrue(isPlatformSocketImpl(si));
+
+        // check if SocketImpl.fd is set
+        Field f = SocketImpl.class.getDeclaredField("fd");
+        f.setAccessible(true);
+        FileDescriptor fd = (FileDescriptor) f.get(si);
+        if (fd == null) {
+            return false;  // not created
+        }
+
+        // call getSocket to get the value of the SO_REUSEADDR socket option
+        Method m = SocketImpl.class.getDeclaredMethod("getOption", SocketOption.class);
+        m.setAccessible(true);
+        try {
+            m.invoke(si, StandardSocketOptions.SO_REUSEADDR);
```

Comment:
If I understand this correctly, we are calling this method just to have the underlying native call to go through and the actual choice of the option name (in this case the SO_REUSEADDR) isn't of any importance?

Reply:
That's right, legacy SocketImpl doesn't define an API to test if has an open socket.